res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,401 @@
+        }
+
+        __onEvent(evt) {
+            if (evt === undefined) {
+                this.__timer.cancel();
+
+            } else if (evt in this.__events) {
+                if (this.__timer.settled) {
+                    this.__timer = new script.TimerPromise(this.__windowDuration, true).then(() => {
+                        this.__dispatch(this.__events);
+                        this.__reset();
+                    });
+                }
+                this.__events[evt]++;
+            }
+        }
+
+        __reset() {
+            this.__events = this.__eventNames.reduce((obj, name) => {
+                obj[name] = 0;
+                return obj;
+            }, {});
+        }
+    }
+
+    /** @typedef {function():void} ButtonActionCallback */
+
+    /**
+     * @callback ButtonEventCallback
+     * @param {Button.Events} event
+     */
+
+    /**
+     * @typedef {ComponentOpts} ButtonOpts
+     * @property {Button.Types} [type=Button.Types.PUSH]
+     * @property {number} [on=127]
+     * @property {number} [off=0]
+     * @property {number} [longPressTimeout=275] Time millisecondsto distinguish a short press
+     *      from a long press. It is recommended to refer to it (as this.longPressTimeout)
+     *      in any Buttons that act differently with short and long presses
+     *      to keep the timeouts uniform
+     * @property {ButtonEventCallback} [onEvent=undefined]
+     * @property {ButtonActionCallback} [onPress=undefined]
+     * @property {ButtonActionCallback} [onLongPress=undefined]
+     * @property {ButtonActionCallback} [onDoublePress=undefined]
+     */
+    class Button extends Component {
+        /** @param {ButtonOpts} opts */
+        constructor({
+            type = Button.Types.PUSH,
+            on = 127,
+            off = 0,
+            longPressTimeout = 275,
+            onEvent = undefined,
+            onPress = undefined,
+            onLongPress = undefined,
+            onDoublePress = undefined
+        } = {}) {
+            super(arguments[0]);
+
+            this.type = type;
+            this.on = on;
+            this.off = off;
+            this.longPressTimeout = longPressTimeout;
+
+            // Replacing default methods by user specified ones
+            if (onEvent instanceof Function) {
+                this.onEvent = onEvent.bind(this);
+            }
+
+            if (onPress instanceof Function) {
+                this.onPress = onPress.bind(this);
+            }
+
+            if (onLongPress instanceof Function) {
+                this.onLongPress = onLongPress.bind(this);
+            }
+
+            if (onDoublePress instanceof Function) {
+                this.onDoublePress = onDoublePress.bind(this);
+            }
+
+            this.__btnEvts = new InputTimeWindow(this.longPressTimeout, Object.values(Button.Events), evts => {
+                if (evts[Button.Events.PRESS] === 1 && evts[Button.Events.RELEASE] === 1) {
+                    this.onPress();
```

Comment:
I may be misinterpreting the code, but from what I could tell this abstraction would only call the callback after it has collected all the events for `this.longPressTimeout` milliseconds. That means that even simple button presses where you just want to run code once the button has been pressed (not even waiting for the release) would introduce this huge artificial latency. This is not an option, input latency has to be reduced to a sensible minimum (essentially as much as possible, though I wouldn't sacrifice other aspects such as code quality for it).

Reply:
Yes, I agree. Which is why [this code don't execute if the `onLongPress` and `onDoublePress` methods are not overrided](https://github.com/mixxxdj/mixxx/pull/13437/files#diff-7e35655817545d6748ef8999d6bfb61a8ae3b1b8c4fdedf9977bbb23e3f9a949R343).

Sadly, as of now, it's not possible to bind several scripts to the same (midiStatus, midiNo) tuple. Which means you must use the same component if you want to use several interaction to the same button. I think it's an acceptable tradeoff.

Or we could modify the engine to execute multiple callbacks for the same midi signal.

Reply:
Ok, reflecting about this, I think I may have a more reactive algorithm for this feature. It goes like this:
- the event bus recieves a push event:
  - if the event is the first detected, wait 50ms
  - after 50 ms,
    - if the bus received a release event, we're either on a single or double push, wait another 50ms,
        - if the bus didn't receive a second push event, we're on a single push, call the dispatch callback signaling a single push, reset the bus, end of cycle, reaction time: 100ms.
        - if the bus received a second push event, we're on a double push, call the , reset the bus, end of cycle, reaction time: 100ms.
    - if the bus didn't receive a release event, wait until the end of `longPressTimeout` and call the dispatch callback signaling a long push.
    - if a release event is received in between, then it's a single push.

Would that algorithm be more acceptable?